src/core/unit.c

```diff
@@ -5394,29 +5396,31 @@ int unit_prepare_exec(Unit *u) {
         return 0;
 }
 
-static void log_leftover(pid_t pid, int sig, void *userdata) {
+static bool log_leftover(pid_t pid, int sig, void *userdata) {
         _cleanup_free_ char *comm = NULL;
 
         (void) get_process_comm(pid, &comm);
 
         if (comm && comm[0] == '(') /* Most likely our own helper process (PAM?), ignore */
-                return;
+                return false;
 
-        log_unit_warning(userdata,
+        log_unit_warning((Unit *) userdata,
                          "Found left-over process " PID_FMT " (%s) in control group while starting unit. Ignoring.\n"
                          "This usually indicates unclean termination of a previous run, or service implementation deficiencies.",
                          pid, strna(comm));
+
+        return true;
 }
 
-void unit_warn_leftover_processes(Unit *u) {
+bool unit_warn_leftover_processes(Unit *u) {
         assert(u);
 
         (void) unit_pick_cgroup_path(u);
 
         if (!u->cgroup_path)
-                return;
+                return false;
 
-        (void) cg_kill_recursive(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, 0, 0, NULL, log_leftover, u);
+        return cg_kill_recursive(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, 0, 0, NULL, log_leftover, u) == 1;
```

Comment:
i think we should propagate the return as is (i.e. change the return type of unit_warn_leftover_processes() to "int"), and let the caller eat up any errors. It's more in line with the vast majority of other functions we have, and I think more generic, if we propagate errors always and eat them up as far up the call chain as appropriate but not needlessly far down.

i.e. make this function return "int", so that it returns < 0 on error, == 0 if no offending processes were found and > 0 if there were.